File: src/planet.rs
use raylib::{Vector2, RaylibHandle, Color, consts};

pub const PLANET_DENSITY: f32 = 5000.0; // kg m^-2
const TRAIL_PLACEMENT_INTERVAL: f32 = 0.04;
const TRAIL_LIFESPAN: f32 = 1.0;

pub struct TrailNode {
	pub pos: Vector2,
	pub time_created: f32
}

pub struct Planet {
	pub id: u32,
	pub pos: Vector2,
	pub vel: Vector2,
	pub radius: f32,
	pub mass: f32,
	pub res_force: Vector2,  // Resultant force
	pub stationary: bool,
	pub is_prediction: bool,

	pub trail_nodes: Vec<TrailNode>,
	pub trail_timer: f32
}

impl Default for Planet {
	fn default() -> Planet {
		Planet {
			id: 0,
			pos: Vector2::zero(),
			vel: Vector2::zero(),
			radius: 5.0,
			mass: Planet::get_mass(2.0),
			res_force: Vector2::zero(),
			stationary: false,
			is_prediction: false,
			trail_nodes: vec![],
			trail_timer: 0.0,
		}
	}
}

impl PartialEq for Planet {
	fn eq(&self, other: &Planet) -> bool {
		self.id == other.id
	}
}

impl Eq for Planet {}

impl Planet {
	pub fn new(id_num: u32, p: Vector2, v: Vector2, r: f32, ma: f32, stat: bool) -> Planet {
		Planet {
			id: id_num,
			pos: p,
			vel: v,
			radius: r,
			mass: if ma == 0.0 { Planet::get_mass(r) } else { ma },
			stationary: stat,
			is_prediction: false,
			..Default::default()
		}
	}

	pub fn draw(&self, rl: &RaylibHandle, time: f32, trails: bool) {
		if trails {
			self.draw_trail(rl, time);
		}
		let col = if self.is_prediction { Color::RED } else { Color::RAYWHITE };
		
		rl.draw_circle_v(self.pos, self.radius, col);
		//self.draw_debug(rl);
	}

	pub fn update(&mut self, dt: f32, time: f32, trails: bool) {
		if !self.is_prediction {
			self.kill_dead_trails(time);
		}

		if !self.stationary {
			if trails {
				self.trail_timer += dt;
				if self.trail_timer >= TRAIL_PLACEMENT_INTERVAL {
					self.place_trail(time);
					self.trail_timer = 0.0;
				}

				let len = self.trail_nodes.len();
				if len > 0 {
					self.trail_nodes[len-1].pos = self.pos;
				}
			}

			self.vel += self.res_force.scale_by(dt/self.mass);  // F = ma, a = F / m, a * dt = vel increase, f * dt/mass = vel change
			self.res_force = Vector2::zero();
		
			self.pos += self.vel.scale_by(dt);
		}
	}

	fn draw_trail(&self, rl: &RaylibHandle, time: f32) {
		let mut col = if self.is_prediction { Color::RED } else { Color::BLUE };
		for i in 0..self.trail_nodes.len() {
			if i > 0 {
				if !self.is_prediction {
					col.a = ((1.0 - ((time - self.trail_nodes[i].time_created)/TRAIL_LIFESPAN)).powi(2) * 255.0).min(255.0) as u8;
				}
				rl.draw_line_ex(self.trail_nodes[i-1].pos, self.trail_nodes[i].pos, 2.0, col);
			}
		}
	}

	#[inline]
	fn place_trail(&mut self, time: f32) {
		//if self.vel.length() > self.radius/2.0 {
		self.trail_nodes.push(TrailNode { pos: self.pos, time_created: time });
		//}
	}

	#[inline]
	fn kill_dead_trails(&mut self, time: f32) {
		self.trail_nodes.retain(|t|time - t.time_created <= TRAIL_LIFESPAN);
	}

	#[inline]
	fn get_mass(radius: f32) -> f32 {
		4.0/3.0 * consts::PI as f32 * radius.powi(3) * PLANET_DENSITY
	}

	#[inline] 
	#[allow(dead_code)]
	fn draw_debug(&self, rl: &RaylibHandle) {
		rl.draw_line_ex(self.pos, self.vel + self.pos, 2.0, Color::LIME);
	}

	#[inline]
	pub fn get_momentum(&self) -> Vector2 {
		self.vel.scale_by(self.mass)
	}
}

File: src/planet/planet.go
package planet

import (
	"github.com/gen2brain/raylib-go/raylib"
	"github.com/gen2brain/raylib-go/raymath"
)

const (
	PLNT_DENSITY = 5000
	TRAIL_PLACEMENT_INTERVAL = 0.2
	TRAIL_NODE_DURATION = 2.0
)

type Planet struct {
	ID int32  // max of 2147483647
	Pos rl.Vector3
	Vel rl.Vector3
	ResForce rl.Vector3
	Radius float32
	Mass float32
	trail []*TrailNode
	trailTimer float32
}

func NewPlanet(ID int32, pos, vel rl.Vector3, rad float32) *Planet {
	return &Planet {
		ID: ID,
		Pos: pos,
		Vel: vel,
		ResForce: rl.NewVector3(0, 0, 0),
		Radius: rad,
		Mass: GetMass(rad, PLNT_DENSITY),
	}
}

func (p *Planet) Draw(col rl.Color) {
	rl.DrawSphere(p.Pos, p.Radius, col)
	p.drawTrail()
}

func (p *Planet) Update(dt, time float32) {
	p.killTrailNodes(time)

	raymath.Vector3Scale(&p.ResForce, dt/p.Mass)
	p.Vel = raymath.Vector3Add(p.Vel, p.ResForce)    // F = ma, a = F / m, a * dt = vel increase, f * dt/mass = vel change
	p.ResForce = rl.NewVector3(0, 0, 0)

	p.Pos.X += p.Vel.X * dt;
	p.Pos.Y += p.Vel.Y * dt;
	p.Pos.Z += p.Vel.Z * dt;

	p.trailTimer += dt
	if p.trailTimer >= TRAIL_PLACEMENT_INTERVAL {
		p.placeTrailNode(time)
	}
}

func (p *Planet) drawTrail() {
	for i := 1; i < len(p.trail); i++ {
		col := rl.Blue
		col.A = 255 - uint8(((rl.GetTime() - p.trail[i].TimeCreated) / TRAIL_NODE_DURATION) * 255)
		rl.DrawLine3D(p.trail[i-1].Pos, p.trail[i].Pos, col)
	}
}

func (p *Planet) placeTrailNode(time float32) {
	p.trail = append(p.trail, &TrailNode {
		Pos: p.Pos,
		TimeCreated: time,
	})
}

func (p *Planet) killTrailNodes(time float32) {
	for i := 0; i < len(p.trail); i++ {
		if time - p.trail[i].TimeCreated >= TRAIL_NODE_DURATION {
			p.removeTrailNode(i)
		}
	}
}

func (p *Planet) removeTrailNode(i int) {
	copy(p.trail[i:], p.trail[i+1:])
	p.trail[len(p.trail)-1] = nil
	p.trail = p.trail[:len(p.trail)-1]
}

func (p *Planet) GetSpeed() float32 {
	return raymath.Vector3Length(p.Vel)
}

func (p *Planet) GetMomentum() rl.Vector3 {
	momentum := p.Vel
	raymath.Vector3Scale(&momentum, p.Mass)
	return momentum
}


func GetMass(rad float32, density float32) float32 {
	return ((4.0 * rl.Pi * (rad * rad * rad))/3.0) * density
}


type TrailNode struct {
	Pos rl.Vector3
	TimeCreated float32
}
